[stable/concourse] Upgrade to Concourse 4

wendorf opened the pull request:
**What this PR does / why we need it**:

Supports the recently-released Concourse 4, and drops support for versions earlier than 4.0.0

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #6873

**Special notes for your reviewer**:

This is a big PR, potentially too big. I've split up the pieces of work into individual commits, so can easily remove pieces if needed. I've tried to capture all the desires in #6873, including removing old parameters not supported in 4.0. I have left one deprecated parameter, `CONCOURSE_POSTGRES_DATA_SOURCE`, because it's a little painful to change all of the documentation and usage to use the new split-up parameters, and this was already becoming a big PR. If that's something we would like, though, I can certainly do the work and update this PR.

wendorf committed:
[stable/concourse] Breaking: *.env key to *.extraEnv

Signed-off-by: Dan Wendorf <github@danwendorf.com>

wendorf committed:
[stable/concourse] Drop special-casing for < 3.10

Signed-off-by: Dan Wendorf <github@danwendorf.com>

wendorf committed:
[stable/concourse] Breaking: Concourse 4.0 auth

This is a breaking change to the way authentication is configured. To
support Concourse 4.0's new auth model, the chart requires breaking
changes.

Note:
The "concourse.secret.required" helper is modified to support nested
secret keys. The go templates don't support variadic parameters using
the `...` syntax, so we can't have variable key path parts.

Signed-off-by: Dan Wendorf <github@danwendorf.com>

wendorf committed:
[stable/concourse] Add CloudFoundry authentication

Signed-off-by: Dan Wendorf <github@danwendorf.com>

wendorf committed:
[stable/concourse] Add LDAP authentication

Signed-off-by: Dan Wendorf <github@danwendorf.com>

wendorf committed:
[stable/concourse] Add OIDC authentication

Signed-off-by: Dan Wendorf <github@danwendorf.com>

wendorf committed:
[stable/concourse] Move flags to env vars

We'd like to be consistent with all configuration. There are a couple
cases that are still flags, however, the `--aws-ssm-region` and
`--aws-secretsmanager-region` params can both only be specified by
the $AWS_REGION environment variable, so we use flags to support
providing different values.

To accomplish changing the `--vault-auth-param` flags, we set
`CONCOURSE_VAULT_AUTH_PARAM` to something perhaps a bit confusing:

1) Concourse env vars split on commas to build a list, which is how we
   specify multiple auth params
2) Kubernetes env vars are evaluated in order, so we can refer to
   a previous var with $DOLLAR_SYNTAX to build a "complex" value from
   a simple secret.

Signed-off-by: Dan Wendorf <github@danwendorf.com>

wendorf committed:
[stable/concourse] Remove env vars that don't do anything

There are some params that were in the chart that Concourse doesn't
respect. This is a cleanup commit.

Removed vars:
- CONCOURSE_ALLOW_SELF_SIGNED_CERTIFICATES
- CONCOURSE_OLD_RESOURCE_GRACE_PERIOD
- CONCOURSE_RESOURCE_CACHE_CLEANUP_INTERVAL

Signed-off-by: Dan Wendorf <github@danwendorf.com>

wendorf committed:
[stable/concourse] Major version bump (chart and Concourse)

Bump chart 1.15.0 -> 2.0.0
Bump Concourse 3.14.1 -> 4.1.0

Signed-off-by: Dan Wendorf <github@danwendorf.com>

regisbelson commented on stable/concourse/templates/web-deployment.yaml:
```diff
@@ -319,6 +494,10 @@ spec:
               value: {{ .Values.web.metrics.datadog.agentPort | quote }}
+            {{- if .Values.web.metrics.datadog.prefix }}
+            - name: CONCOURSE_DATADOG_PREFIX
+            - value: {{ .Values.web.metrics.datadog.prefix }}
```
There's an extra `-` in front of `value` here, it breaks the chart when you use Datadog

wendorf replied:
Thanks for the catch! Fixed.

romankartgh commented on stable/concourse/README.md:
```diff
@@ -234,6 +236,15 @@ The following table lists the configurable parameters of the Concourse chart and
+| `secrets.oidcAuth.clientId` | OIDC Authentication: (Required if OIDC enabled) Client id | `nil` |
+| `secrets.oidcAuth.clientSecret` | OIDC Authentication: (Required if OIDC enabled) Client secret | `nil` |
+| `secrets.oidcAuth.scope` | OIDC Authentication: Any additional scopes that need to be requested during authorization | `nil` |
+| `secrets.oidcAuth.groupsKey` | OIDC Authentication: The groups key indicates which claim to use to map external groups to Concourse teams. | `nil` |
```
Would it be better to avoid adding this parameter all together? Turns out OpenID provider in coreos/dex doesn't support group claims: https://github.com/coreos/dex. We integrated with Azure, but there is to way to retrieve the list of groups user belongs to. So this parameter doesn't seem to have any effect on the behavior.

wendorf replied:
I haven't actually used OIDC Authentication at all, so I'm not sure—I've just folded in all of the flags Concourse supports. I recommend asking the Concourse team directly, probably at https://discuss.concourse-ci.org/ .

romankartgh replied:
I filed https://github.com/concourse/concourse/issues/2538, let's see what the response is

romankartgh replied:
sounds like people got this working with okta but not azure. this seems to be okay.
you can disregard my original comment, dan. thanks!

romankartgh commented on stable/concourse/values.yaml:
```diff
@@ -900,6 +906,36 @@ secrets:
+    # scope:
+
+    ## The groups key indicates which claim to use to map external groups to Concourse teams.
+    # groupsKey:
```
Same comment as above

romankartgh reviewed (approved):
we tested this on our side. migrated all parameters from 3.14.1 to 4.0.0. introduced new auth parameters - they get propagated fine for 4.0.0.

william-tran commented on stable/concourse/templates/web-deployment.yaml:
```diff
@@ -30,15 +30,6 @@ spec:
-            {{- end }}
             {{- if .Values.credentialManager.ssm.enabled }}
             - "--aws-ssm-region"
             - "$(CONCOURSE_AWS_SSM_REGION)"
```
Could we get rid of these cmd line args and use the env var `CONCOURSE_AWS_REGION` instead? From the output of `concourse web --help` it looks like that env var is used in both AWS SSM and Secretsmanager cases. Other than that we're looking good!

wendorf replied:
According to the binary I have, the variable it wants to use is `AWS_REGION` (not `CONCOURSE_AWS_REGION`, which would be a little nicer), which seemed like a "heavy" env var to set if you want to configure either the `--aws-secretsmanager-region` or `--aws-ssm-region`. It would mean that configuring one implicitly (and perhaps surprisingly) configures the other one. This is "expected behavior" from Concourse's perspective, but it felt weird to support it, which is why I left it.

If my argument isn't convincing, just let me know and I'll change it :)

william-tran replied:
@wendorf you're right it's `AWS_REGION`, and more importantly, it should be avoided due to this unfortunate issue: https://github.com/concourse/concourse/issues/2191. Lets stick with the cmd line args, but would it be possible to pass the value directly into the args rather than using an env var to propagate it?

wendorf replied:
Sounds good. Just pushed the change.

wendorf committed:
[stable/concourse] AWS_REGION flags dont use env vars

`--aws-ssm-region` and `--aws-secretsmanager-region` don't have unique
environment variable equivalents, so we must pass these flags directly.
Rather than indirecting through unofficial environment variables,
however, we just set the flags directly.

Signed-off-by: Dan Wendorf <github@danwendorf.com>

wendorf committed:
[stable/concourse] Breaking: Individual DB env vars

Use the individual CONCOURSE_POSTGRES_* environment variables rather
than building or expecting a full connection string in
CONCOURSE_POSTGRES_DATA_SOURCE.  CONCOURSE_POSTGRES_DATA_SOURCE has
been deprecated since Concourse 2.7.2, was unintentionally broken
in Concourse 4.1.0, and will be removed instead of fixed
(see https://github.com/concourse/concourse/issues/2549).

The chart does not support the environment variables for specifying
custom certs and keys (CONCOURSE_POSTGRES_CA_CERT,
CONCOURSE_POSTGRES_CLIENT_CERT, and CONCOURSE_POSTGRES_CLIENT_KEY) because they
require additional effort to support, and were not supported in the previous
versions of this chart.

Signed-off-by: Dan Wendorf <dan.wendorf@opendoor.com>
